packages/belos/epetra/test/MINRES/test_pminres_hb.cpp

```diff
@@ -0,0 +1,255 @@
+    // please check the documentation
+    std::string PrecType = "ICT"; // incomplete Cholesky
+    int OverlapLevel = 0; // must be >= 0. If Comm.NumProc() == 1,
+    // it is ignored.
+
+    RCP<Ifpack_Preconditioner> Prec = Teuchos::rcp( Factory.Create(PrecType, &*A, OverlapLevel) );
+    assert(Prec != Teuchos::null);
+
+    // specify parameters for ICT
+    ifpackList.set("fact: drop tolerance", 1e-9);
+    ifpackList.set("fact: ict level-of-fill", 1.0);
+    // the combine mode is on the following:
+    // "Add", "Zero", "Insert", "InsertAdd", "Average", "AbsMax"
+    // Their meaning is as defined in file Epetra_CombineMode.h
+    ifpackList.set("schwarz: combine mode", "Add");
+    // sets the parameters
+    IFPACK_CHK_ERR(Prec->SetParameters(ifpackList));
+
+    // initialize the preconditioner. At this point the matrix must
+    // have been FillComplete()'d, but actual values are ignored.
+    IFPACK_CHK_ERR(Prec->Initialize());
+
+    // Builds the preconditioners, by looking for the values of
+    // the matrix.
+    IFPACK_CHK_ERR(Prec->Compute());
+
+    // Create the Belos preconditioned operator from the Ifpack preconditioner.
+    // NOTE:  This is necessary because Belos expects an operator to apply the
+    //        preconditioner with Apply() NOT ApplyInverse().
+    RCP<Belos::EpetraPrecOp> belosPrec = rcp( new Belos::EpetraPrecOp( Prec ) );
+
+    //
+    // *****Create parameter list for the Minres solver manager*****
+    //
+    const int NumGlobalElements = B->GlobalLength();
+    if (maxiters == -1)
+      maxiters = NumGlobalElements - 1; // maximum number of iterations to run
+    //
+    ParameterList belosList;
+    belosList.set( "Maximum Iterations", maxiters );       // Maximum number of iterations allowed
+    belosList.set( "Convergence Tolerance", tol );         // Relative convergence tolerance requested
+    if (verbose) {
+      belosList.set( "Verbosity", Belos::Errors + Belos::Warnings +
+          Belos::TimingDetails + Belos::FinalSummary + Belos::StatusTestDetails );
+      if (frequency > 0)
+        belosList.set( "Output Frequency", frequency );
+    }
+    else
+      belosList.set( "Verbosity", Belos::Errors + Belos::Warnings );
+    //
+    // *******Construct a preconditioned linear problem********
+    //
+    RCP<Belos::LinearProblem<double,MV,OP> > problem
+      = rcp( new Belos::LinearProblem<double,MV,OP>( A, X, B ) );
+    if (leftprec) {
+      problem->setLeftPrec( belosPrec );
+    }
+    else {
+      problem->setRightPrec( belosPrec );
+    }
+
+    bool set = problem->setProblem();
+    if (set == false) {
+      if (proc_verbose)
+        std::cout << std::endl << "ERROR:  Belos::LinearProblem failed to set up correctly!" << std::endl;
+      return -1;
+    }
+
+    // Create an iterative solver manager.
+    RCP< Belos::SolverManager<double,MV,OP> > solver
+      = rcp( new Belos::MinresSolMgr<double,MV,OP>(problem, rcp(&belosList,false)) );
```

Comment:
There are lots of lines like this in the solver tests. Cleaner way is to use rcpFromRef.

Reply:
Sure, this driver was copied from the preconditioned CG tests, so it's an artifact of copying older code.  I agree there are cleaner ways of handling this.